substratevm/src/com.oracle.svm.hosted.jdk11/src/com/oracle/svm/hosted/ModuleLayerFeature.java

```diff
@@ -0,0 +1,185 @@
+ * If a type is reachable, its module is also reachable and therefore should be included in the
+ * runtime boot module layer.
+ * </p>
+ * <p>
+ * The configuration for the runtime boot module layer is resolved using the module reachability
+ * data provided to us by the analysis as resolve roots.
+ * </p>
+ * <p>
+ * We are purposefully avoiding public API for module layer creation, such as
+ * {@link ModuleLayer#defineModulesWithOneLoader(Configuration, ClassLoader)}, because as a side
+ * effect this will create a new class loader. Instead, we use a private constructor to construct
+ * the {@link ModuleLayer} instance, which we then patch using the module reachability data
+ * provided to us by the analysis.
+ * </p>
+ * <p>
+ * Because the result of this feature is dependant on the analysis results, and because this feature
+ * will add reachable object(s) as it's result, it is necessary to perform the logic during the
+ * analysis, for lack of a better option (even though only the last analysis cycle is sufficient,
+ * but that cannot be known in advance).
+ * </p>
+ */
+@AutomaticFeature
+@Platforms(Platform.HOSTED_ONLY.class)
+public final class ModuleLayerFeature implements Feature {
+
+    private Field moduleNameToModuleField;
+    private Field moduleParentsField;
+    private Constructor<ModuleLayer> moduleLayerConstructor;
+
+    @Override
+    public boolean isInConfiguration(IsInConfigurationAccess access) {
+        return new JDK11OrLater().getAsBoolean();
+    }
+
+    @Override
+    public void afterRegistration(AfterRegistrationAccess access) {
+        ImageSingletons.add(BootModuleLayerSupport.class, new BootModuleLayerSupport());
+        moduleNameToModuleField = ReflectionUtil.lookupField(ModuleLayer.class, "nameToModule");
+        moduleParentsField = ReflectionUtil.lookupField(ModuleLayer.class, "parents");
+        moduleLayerConstructor = ReflectionUtil.lookupConstructor(ModuleLayer.class, Configuration.class, List.class, Function.class);
+    }
+
+    @Override
+    public void beforeAnalysis(BeforeAnalysisAccess access) {
+        access.registerReachabilityHandler(
+                a -> a.registerAsUnsafeAccessed(ReflectionUtil.lookupField(ClassLoader.class, "classLoaderValueMap")),
+                ReflectionUtil.lookupMethod(ClassLoader.class, "trySetObjectField", String.class, Object.class)
+        );
+        RuntimeReflection.register(ReflectionUtil.lookupField(ClassLoader.class, "classLoaderValueMap"));
+    }
+
+    @Override
+    public void duringAnalysis(DuringAnalysisAccess access) {
+        FeatureImpl.DuringAnalysisAccessImpl accessImpl = (FeatureImpl.DuringAnalysisAccessImpl) access;
+        AnalysisUniverse universe = accessImpl.getUniverse();
+
+        Map<String, Module> reachableModules = universe.getTypes()
+                .stream()
+                .filter(t -> t.isReachable() && !t.isArray())
+                .map(t -> t.getJavaClass().getModule())
```

Comment:
Probably a good idea to add
```
.distinct()
```
after
```
.map(t -> t.getJavaClass().getModule())
```
(no need to look at any prev. looked module)

Not sure if it makes a measurable difference though (trading space for avoiding redundant recomputations).

Reply:
Done. It is also a question if `distinct()` will perform more efficiently compared to `toMap` collector because `distinct()` can work in O(n) with a hash map as a helper data structure. In this case the map is actually the end product as well as a helper so it seemed natural to remove duplicates in that place. It is an interesting question and something to think about if we really wish this part of code to be as efficient as possible, I am not sure what is the best approach.